pulser-simulation/pulser_simulation/qutip_result.py

```diff
@@ -126,33 +126,84 @@ def _weights(self) -> np.ndarray:
+    ) -> List[int]:
+        """Finds the subsystems to keep based on the target basis.
+
+        Args:
+            to_basis: The basis to which the state should be transformed.
+
+        Returns:
+            A list of indices representing the relevant subsystems.
+        """
+        if to_basis == "ground-rydberg":
+            ex_state = "2"
+        elif to_basis == "digital":
+            ex_state = "0"
+        else:
+            raise ValueError(
+                "'to_basis' must be 'ground-rydberg' or 'digital'."
+            )
+
+        # Find the subsystems to keep.
+        ex_inds = [
+            i
+            for i in range(3**self._size)
+            if ex_state in np.base_repr(i, base=3).zfill(self._size)
+        ]
+
+        ex_probs = np.abs(self.state.extract_states(ex_inds).full()) ** 2
+        if not np.all(np.isclose(ex_probs, 0, atol=tol)):
+            raise TypeError(
+                "Can't reduce to chosen basis because the population of a "
+                "state to eliminate is above the allowed tolerance."
+            )
+
+        return ex_inds
+
+    def _reduce_state_vector_to_basis(
+        self, ex_inds: List[int], normalize: bool
     ) -> qutip.Qobj:
-        """Gets the state with some optional post-processing.
+        """Reduces the state vector by eliminating the states at given indices.
 
         Args:
-            reduce_to_basis: Reduces the full state vector
-                to the given basis ("ground-rydberg" or "digital"), if the
-                population of the states to be ignored is negligible. Doesn't
-                apply to XY mode.
-            ignore_global_phase: If True and if the final state is a vector,
-                changes the final state's global phase such that the largest
-                term (in absolute value) is real.
-            tol: Maximum allowed population of each eliminated state.
-            normalize: Whether to normalize the reduced state.
+            ex_inds: The indices of the states to eliminate.
 
         Returns:
-            The resulting state.
+            A Qobj instance representing the state vector with the states eliminated.
+        """
+        state = self.state.copy()
+        state = state.eliminate_states(ex_inds, normalize=normalize)
+        return state
 
-        Raises:
-            TypeError: If trying to reduce to a basis that would eliminate
-                states with significant occupation probabilites.
+    def _reduce_density_matrix_to_basis(
+        self, ex_inds: List[int], normalize: bool
+    ) -> qutip.Qobj:
+        """Reduces the density matrix by tracing out the subsystems at given indices.
+
+        Args:
+            ex_inds: The indices of the subsystems to trace out.
+
+        Returns:
+            The subsystem state, reduced by tracing out the subsystems at given indices.
         """
+        state = self.state.copy()
+        reduced_state = state.ptrace(ex_inds)
```

Comment:
Two things here:
1. I think that `ptrace()` **keeps** the provided indices, rather than exclude them.
2. This does not have the same behaviour as `eliminate_states()`. The partial trace removes part of a multipartite system (eg if you have a two-qubit system and want the state of only one) while `eliminate_states()` will remove specifc indices (eg turning a qutrit into a qubit). Also, it seems that you can use `eliminate_states()` on density matrices too, so there's no need to use the partial trace here.